dpcpp/components/bitvector.dp.hpp

```diff
@@ -0,0 +1,114 @@
+// SPDX-FileCopyrightText: 2017 - 2025 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_DPCPP_COMPONENTS_BITVECTOR_DP_HPP_
+#define GKO_DPCPP_COMPONENTS_BITVECTOR_DP_HPP_
+
+
+#include <sycl/sycl.hpp>
+
+#include <ginkgo/core/base/intrinsics.hpp>
+
+#include "core/components/bitvector.hpp"
+#include "core/components/fill_array_kernels.hpp"
+#include "core/components/prefix_sum_kernels.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace GKO_DEVICE_NAMESPACE {
+namespace bitvector {
+
+
+template <typename IndexType, typename DevicePredicate>
+gko::bitvector<IndexType> from_predicate(
+    std::shared_ptr<const DefaultExecutor> exec, IndexType size,
+    DevicePredicate device_predicate)
+{
+    using storage_type = typename device_bitvector<IndexType>::storage_type;
+    constexpr auto block_size = device_bitvector<IndexType>::block_size;
+    const auto num_blocks = static_cast<size_type>(ceildiv(size, block_size));
+    array<uint32> bit_array{exec, num_blocks};
+    array<IndexType> rank_array{exec, num_blocks};
+    const auto bits = bit_array.get_data();
+    const auto ranks = rank_array.get_data();
+    const auto queue = exec->get_queue();
+    queue->submit([&](sycl::handler& cgh) {
+        cgh.parallel_for(num_blocks, [=](sycl::id<1> block_i) {
+            const auto base_i = static_cast<IndexType>(block_i) * block_size;
+            storage_type mask{};
+            const auto local_op = [&](int local_i) {
+                const storage_type bit =
+                    device_predicate(base_i + local_i) ? 1 : 0;
+                mask |= bit << local_i;
+            };
+            if (base_i + block_size <= size) {
+                for (int local_i = 0; local_i < block_size; local_i++) {
+                    local_op(local_i);
+                }
```

Comment:
```suggestion
#pragma unroll
                for (int local_i = 0; local_i < block_size; local_i++) {
                    local_op(local_i);
                }
```
to make them indeed different from known size.
the later one can be still merge into this unroll loop by wrapping local_op into condition.
the #check should be the same as the current one.
```
for (...) {
  if (base_i + local_i < size) {
    local_op(local_i);
  }
}
```